File: C++/프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include <string>  

using namespace std;

bool solution(string s) {
    int count = 0;

    for(int i = 0; i < s.size(); i++) {
        if(s[i] == '(') {
            count++;  // 열린 괄호를 만나면 카운트 증가
        } else {
            count--;  // 닫힌 괄호를 만나면 카운트 감소
        }

        // 만약 카운트가 음수가 되면 올바르지 않은 괄호
        if(count < 0) {
            return false;  
        }
    }

    bool answer = true;
    if (count != 0) {
        answer = false;
    }
    return answer;
}

File: Java/프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
class Solution {
    boolean solution(String s) {
        int cnt = 0;  // 여는 괄호와 닫는 괄호의 균형을 맞추기 위한 변수

        for (int i = 0; i < s.length(); i++) {
            // 여는 괄호면 증가
            if (s.charAt(i) == '(') {
                cnt++;
            } 
            // 닫는 괄호면 감소
            else {
                cnt--;
            }
            
            // 만약 중간에 닫는 괄호가 먼저 나온 경우 (cnt가 음수가 되면) 바로 false 반환
            if (cnt < 0) {
                return false;
            }
        }
        
        // 모든 문자를 확인 후 여는 괄호와 닫는 괄호의 수가 같으면 true, 아니면 false
        return cnt == 0;
    }
}
